Output from previous steps not available in following steps

I have a workflow with a form input, a first SQL query using the form input, and then a second SQL query using output of the first step.

<img width="1724" height="1111" alt="Image" src="https://github.com/user-attachments/assets/ec94229d-2df8-4f55-83e5-46582218a547" />

Running it with docker image 0.3.0 works as expected.
But with latest image and worker (0.4.0) I got an error:

```
worker-1  | 2025-11-14 10:45:48.570 | INFO     | autodoc.workflow.workflow:build_contexts:165 - contexts before loading: [{'year': 2025}]
worker-1  | 2025-11-14 10:45:48.573 | INFO     | autodoc.workflow.workflow:build_contexts:169 - contexts after loading: [{'year': 2025, 'schedule_id': 29, 'schedule_name': 'foo', 'coach_name': 'bar'}]
worker-1  | 2025-11-14 10:45:48.574 | INFO     | autodoc.workflow.workflow:build_contexts:144 - processing source_id 12: SQL RecordSet
worker-1  | 2025-11-14 10:45:48.574 | INFO     | autodoc.workflow.workflow:build_contexts:147 - uploaded_filename=None
worker-1  | [2025-11-14 10:45:48,770] [PID 7] [Thread-4] [dramatiq.worker.WorkerThread] [ERROR] Failed to process message process_instance(instance_id=10, upload_mapping={'template': '/upload_dir/template.csv'}, form_data={'year': 2025}) with unhandled exception.
worker-1  | Traceback (most recent call last):
worker-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 1815, in _execute_context
worker-1  |     context = constructor(
worker-1  |               ^^^^^^^^^^^^
worker-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/engine/default.py", line 1373, in _init_compiled
worker-1  |     compiled.construct_params(
worker-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/sql/compiler.py", line 1930, in construct_params
worker-1  |     raise exc.InvalidRequestError(
worker-1  | sqlalchemy.exc.InvalidRequestError: A value is required for bind parameter 'schedule_id' (Background on this error at: https://sqlalche.me/e/20/cd3x)
worker-1  |
worker-1  | The above exception was the direct cause of the following exception:
worker-1  |
worker-1  | Traceback (most recent call last):
worker-1  |   File "/usr/local/lib/python3.12/site-packages/dramatiq/worker.py", line 487, in process_message
worker-1  |     res = actor(*message.args, **message.kwargs)
worker-1  |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
worker-1  |   File "/usr/local/lib/python3.12/site-packages/dramatiq/actor.py", line 185, in __call__
worker-1  |     return self.fn(*args, **kwargs)
worker-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^
worker-1  |   File "/worker/autodoc/tasks.py", line 28, in process_instance
worker-1  |     workflow.process()
worker-1  |   File "/worker/autodoc/workflow/workflow.py", line 91, in process
worker-1  |     self.contexts = self.build_contexts()
worker-1  |                     ^^^^^^^^^^^^^^^^^^^^^
worker-1  |   File "/worker/autodoc/workflow/workflow.py", line 160, in build_contexts
worker-1  |     source_service.load_data(current_data=self.data)
worker-1  |   File "/worker/autodoc/source/sql_source.py", line 78, in load_data
worker-1  |     recordset = self.meta_database.recordset(
worker-1  |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
worker-1  |   File "/worker/autodoc/metasource.py", line 19, in recordset
worker-1  |     headings, data = self.get_data(sql, params)
worker-1  |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^
worker-1  |   File "/worker/autodoc/metasource.py", line 65, in get_data
worker-1  |     result = connection.execute(bound_sql)
worker-1  |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
worker-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 1418, in execute
worker-1  |     return meth(
worker-1  |            ^^^^^
worker-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/sql/elements.py", line 515, in _execute_on_connection
worker-1  |     return connection._execute_clauseelement(
worker-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
worker-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 1640, in _execute_clauseelement
worker-1  |     ret = self._execute_context(
worker-1  |           ^^^^^^^^^^^^^^^^^^^^^^
worker-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 1821, in _execute_context
worker-1  |     self._handle_dbapi_exception(
worker-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 2353, in _handle_dbapi_exception
worker-1  |     raise sqlalchemy_exception.with_traceback(exc_info[2]) from e
worker-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/engine/base.py", line 1815, in _execute_context
worker-1  |     context = constructor(
worker-1  |               ^^^^^^^^^^^^
worker-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/engine/default.py", line 1373, in _init_compiled
worker-1  |     compiled.construct_params(
worker-1  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/sql/compiler.py", line 1930, in construct_params
worker-1  |     raise exc.InvalidRequestError(
worker-1  | sqlalchemy.exc.StatementError: (sqlalchemy.exc.InvalidRequestError) A value is required for bind parameter 'schedule_id'
worker-1  | [SQL: select m.lastname, m.firstname, m.date_of_birth,
worker-1  | mse.description as membership, m.comments
worker-1  | from `Member` m
worker-1  | join MembershipStatusEnum mse on m.membership_status = mse.id
worker-1  | join ScheduleMemberLink_2025 sml on sml.member_id = m.id
worker-1  | where sml.schedule_id = %(schedule_id)s
worker-1  | ORDER BY mse.id desc, lastname, firstname]
worker-1  | (Background on this error at: https://sqlalche.me/e/20/cd3x)
```

--- a/autodoc/source/source.py
+++ b/autodoc/source/source.py
@@ -10,7 +10,7 @@ from autodoc.storage_service import get_storage_service
 class SourceService(ABC):
     """Base Interface for a Source object."""
 
-    # data: dict | list
+    data: dict | list
     # is_multi_record: bool
     # storage_service: StorageService | None
     # path: Path

--- a/autodoc/workflow/workflow.py
+++ b/autodoc/workflow/workflow.py
@@ -122,16 +122,57 @@ class Workflow:
             )
             self.manager.commit()
 
-            source_service.load_data(current_data=self.data)
+            logger.info(f"contexts before loading source {source.Id}: {contexts}")
+
+            """
+            when updating contexts, single record enriches all current
+            contexts, while a multi record will copy for each output and then
+            enrich that context. Note, both multi_record AND splitter must be
+            set for context expansion before enrichment. That is because a
+            multi_record may be writing to a single field rather than
+            expanding.
+
+            single record enrichment [{"Id": 1}] -> [{"Id": 1, "name":
+            "Barry"}] [{"Id": 1}, {"Id": 2}] -> [{"Id": 1, "name": "Barry"},
+            {"Id": 1, "name": "Barry"}]
+
+            multi_record + splitter enrichment [{"something": 1}] ->
+            [{"something": 1, "name": "Barry"}, {"something": 1, "name":
+            "Gary"}]
+            """
+
+            next_contexts = []
+
+            logger.info(f"{len(contexts)} contexts to loop through.")
+
+            for context in contexts:
+                logger.info(f"Running context {context}")
+                source_service.load_data(current_data=context)
+
+                logger.info(f"source has loaded data {source_service.data}")
+
+                if source_service.is_multi_record:
+                    if source_service.source.Splitter:
+
+                        for record in source_service.data:
+                            merged = context.copy()
+                            merged.update(record)
+                            next_contexts.append(merged)
+
+                    else:
+                        merged = context.copy()
+                        merged.update({source_service.source.FieldName: source_service.data})
+                        next_contexts.append(merged)
+
+                else:
+                    merged = context.copy()
+                    merged.update(source_service.data)
+                    next_contexts.append(merged)
+
+            contexts = next_contexts
 
             self.manager.source_instances.set_loaded(source_instance_id=source_instance.Id)
             self.manager.commit()
-
-            logger.info(f"contexts before loading: {contexts}")
-
-            contexts = self.expand(contexts, source_service)
-
-            logger.info(f"contexts after loading: {contexts}")
 
         return contexts
 
@@ -160,7 +201,6 @@ class Workflow:
 
         if not check:
             if reasons:
-
                 reasons_text = "|".join(r for r in reasons if r)
 
             else:

